src/balance.rs

```diff
@@ -26,41 +41,219 @@ impl<T: Config> From<Error<T>> for StatusCode {
         }
     }
 }
 
-impl<T: Config> Default for BalanceAdapter<T> {
+impl<T: Config> Default for BalanceAdapter<T>
+where
+    AccountIdOf<T>: std::cmp::Eq + std::hash::Hash,
+    BalanceOf<T>: From<u128> + Into<u128>,
+{
     fn default() -> Self {
         Self::new()
     }
 }
 
-impl<T: Config> BalanceHandler for BalanceAdapter<T> {
+impl<T: Config> BalanceHandler for BalanceAdapter<T>
+where
+    AccountIdOf<T>: std::cmp::Eq + std::hash::Hash,
+    BalanceOf<T>: From<u128> + Into<u128>,
+{
     type Error = StatusCode;
 
     fn transfer(
         &self,
-        _src: AccountAddress,
-        _dst: AccountAddress,
-        _cheque_amount: u128,
+        src: AccountAddress,
+        dst: AccountAddress,
+        cheque_amount: u128,
     ) -> Result<bool, Self::Error> {
         // TODO(rqnsom): This won't break the MoveVM unless the balance module is accessed.
-        unimplemented!()
+
+        // use move_core_types::vm_status::StatusCode::{ARITHMETIC_ERROR, INSUFFICIENT_BALANCE};
+        use move_core_types::vm_status::StatusCode::INSUFFICIENT_BALANCE;
+
+        let from = Pallet::<T>::to_native_account(&src)?;
+        let to = Pallet::<T>::to_native_account(&dst)?;
+        let amount = BalanceOf::<T>::from(cheque_amount);
+
+        self.ensure_account_exists(&from);
+        self.ensure_account_exists(&to);
+
+        // if *(self.balances.borrow().get(&from).ok_or(ARITHMETIC_ERROR)?) > amount {
+        //     {
+        //         let mut balances = self.balances.borrow_mut();
+        //         *(balances.get_mut(&from).ok_or(ARITHMETIC_ERROR)?) -= amount;
+        //     }
+        //     if let Some(balance_to) = self.balances.borrow_mut().get_mut(&to) {
+        //         *balance_to += amount;
+        //         Ok(true)
+        //     } else {
+        //         let mut balances = self.balances.borrow_mut();
+        //         *(balances.get_mut(&from).ok_or(ARITHMETIC_ERROR)?) += amount;
+        //         Err(ARITHMETIC_ERROR)
+        //     }
+        // } else {
+        //     Err(INSUFFICIENT_BALANCE)
+        // }
+        let mut balances = self.balances.borrow_mut();
+        if *(balances.get(&from).unwrap()) < amount {
+            Err(INSUFFICIENT_BALANCE)
+        } else {
+            *(balances.get_mut(&from).unwrap()) -= amount;
+            *(balances.get_mut(&to).unwrap()) += amount;
+            Ok(true)
+        }
     }
 
-    fn cheque_amount(&self, _account: AccountAddress) -> Result<u128, Self::Error> {
+    fn cheque_amount(&self, account: AccountAddress) -> Result<u128, Self::Error> {
         // TODO(rqnsom): This won't break the MoveVM unless the balance module is accessed.
-        unimplemented!()
+
+        use move_core_types::vm_status::StatusCode::SENDING_ACCOUNT_DOES_NOT_EXIST;
+
+        let acc = Pallet::<T>::to_native_account(&account)?;
+        let balances = self.balances.borrow();
+        let balance = balances.get(&acc).ok_or(SENDING_ACCOUNT_DOES_NOT_EXIST)?;
```

Comment:
Hm, maybe it would be better to return zero in case we don't have a cheque written by some account?

Imagine a user writing a move module like this:
```
// charge an account (which might exist) if the account wrote me a cheque
// mint him something if the transfer didn't fail
```

Reply:
It got updated with the changes to `HashMap<AccountId, Balance>` thing... check again :blush: